### DISCUSSION TITLE
web3modal is failing to detect the MetaMask have installed in the web browser

### DISCUSSION THREAD

**Discussion by javaspeak** [WalletConnect/walletconnect-monorepo#3586] [Developer Support] [answered]
web3modal is failing to detect the MetaMask have installed in the web browser

<img width="384" alt="Screenshot 2023-09-03 at 14 26 40" src="https://github.com/WalletConnect/walletconnect-monorepo/assets/26520180/35db7db4-7021-4438-8256-8ddc83f6c61b">
   
I am using react and my index.tsx looks like:

```
import React from 'react';
import { createRoot } from 'react-dom/client';
import App from './App';
import { Web3Modal } from '@web3modal/react'
import { configureChains, createConfig, WagmiConfig } from 'wagmi';
import { EthereumClient, w3mConnectors, w3mProvider } from '@web3modal/ethereum';
import { arbitrum, mainnet, polygon } from 'wagmi/chains';

const chains = [arbitrum, mainnet, polygon];

// Change this projectId to your one - see instructions in README.md
const projectId = "change_for_your_project_id";

const { publicClient } = configureChains(chains, [w3mProvider({ projectId })])

const wagmiConfig = createConfig({
  autoConnect: true,
  connectors: w3mConnectors({ projectId, chains }),
  publicClient
});

const ethereumClient = new EthereumClient(wagmiConfig, chains)

createRoot(document.getElementById('register-page-root')!).render(
  <React.StrictMode>
    <WagmiConfig config={wagmiConfig}>
      <App />
    </WagmiConfig>
    <Web3Modal projectId={projectId} ethereumClient={ethereumClient} />
  </React.StrictMode>,
);

```
My App.tsx looks like:

```
// @ts-ignore
import { useWeb3Modal } from '@web3modal/react';

function App() {

  const { open } = useWeb3Modal();

  const connectToWalletClick = async () => {
    // This connects your wallet
    await open();
  }

  return (

    <div>
      <button type='button' style={{ cursor: 'pointer' }}
        className={`x-utu-btn x-utu-btn-light border-radius`}
        onClick={connectToWalletClick}>Connect to Wallet</button>
    </div>
  );
};

export default App;

```

My package.json looks like:

```
{
  "name": "react-browser-extension-tab-register",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "start": "react-app-rewired start",
    "test": "react-app-rewired test",
    "rewire-config": "tsc --removeComments config-overrides.ts",
    "build": "react-app-rewired build"
  },
  "dependencies": {
    "@web3modal/ethereum": "^2.7.1",
    "@web3modal/react": "^2.7.1",
    "wagmi": "^1.3.11",
    "web3modal": "^2.7.1"
  },
  "devDependencies": {},
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  }
}
```

I am using:

     Node v18.9.0

Any ideas why Wallet Connect 2 is not detecting my Meta Mask.  

I am using Meta Mask:

     10.35.1

I am using Chrome:

     116.0.5845.140 

I opened a tab from a chrome extension to trigger the wallet connect.  The url of the tab looks like:

    chrome-extension://oahhaapnfpkfnbgeopliehfgenifngde/register.html

Am beginning to think that a browser version of meta mask cannot be accessed from this tab because it somehow cannot see the metamask context.  Am I right here?

Any hints from a smarter more experienced person than me would be much appreciated.  I will owe you a lorry of beers.

Thank you for your time in advance and have a great day!

**Comment by glitch-txs** [comment]
hi, do you have any other wallet extensions in your browser? could you try disabling them all except MetaMask?

**Reply by javaspeak** [reply]
The reason that Meta Mask is not found,  is that Wallet Connect depends on window.ethereum being visible and the Chrome Extension Security model I am developing with does not allow access to  window.ethereum.

MetaMask does have a Port Provider that does not depend on window.ethereum being visible.  However Wallet Connect has not integrated with it.  I think MetaMask brought out the port provider just so chrome extensions can use it.

The internal logic in the Wallet Connect library would be something like:

```
if ( ! window.etherum ) {
     if ( canConnectToPortProviderOfMetaMask ) {
           showAndConnectUsingPortProviderOfMetaMask();
     }
}
```


Then Wallet Connect could urge the other Wallels to also provide a Port Provider and it could do similar checks with them.

**Reply by xzilja** [reply, marked as answer]
I'm afraid this is a very specific use case that we won't be able to support in a near future. We rely on standards like `window.ethereum` to be available. Your best bet would be to fork our repo and add this logic in for your use case.
